pkg/option/resolver/resolver.go

```diff
@@ -62,9 +62,30 @@ func ResolveConfigurations(ctx context.Context, client client.Clientset, nodeNam
 		matchKeys = sets.New(denyConfigKeys...)
 	}
 
+	// Find if there are any KindNodeConfig v2 sources first and prefer that
+	// over v2alpha1 if so.
+	sourcesV2Only := make([]ConfigSource, 0, len(sources))
+	sourcesExceptV2 := make([]ConfigSource, 0, len(sources))
```

Comment:
I don't understand why we have to track this; the apiserver itself will only serve a single version, and it will upconvert any v2alpha1 resources to v2 transparently.

The only reason we have to have a fallback in the first place is that we might beat the operator to the CRD upgrade, and we really really really cannot skip an existing CiliumNodeConfig, whatever the scenario.

Reply:
> we really really really cannot skip an existing CiliumNodeConfig, whatever the scenario.

If that's the case, then wouldn't this entire PR be invalid?

Reply:
Is this conversation resolved?

Reply:
I think we can't until Casey gets back. He's out for now, but we'll wait until he's back.

Reply:
@christarazi I agree the fallback in `readNodeConfigsAllVersions` is needed, but I don't think we need to track sources by version. So AIUI these separate lists are unnecessary. Correct me if there's a case I'm missing.

Reply:
@squeed So the reason for tracking sources by version is that the iteration order of sources matters, so we must put the v2 versions before the v2alpha1 versions. The iteration order matters because `readNodeConfigsAllVersions` assumes that the v2 versions are handled first, and _then_ falls back to v2alpha1.

Reply:
@christarazi it's not possible for a source to be one version and not the other; the apiserver automatically converts between the two versions for resources of the same name.

We just need to know globally whether or not the apiserver supports `v2` versus `v2alpha1`

Reply:
@squeed That makes sense and it's possible that you can have two sources of different names and also different versions (v2 / v2alpha1), right? The code here addresses that case, as shown by the unit tests. If I remove this code, the unit tests fail. If we don't ensure that v2 is handled first then we will continue to get warning logs about deprecation because v2alpha1 will always be fetched as well. In other words, the last states that if v2 is found, v2alpha1 is ignored. I assume this is ok because we have deprecated v2alpha1 so the user should have migrated already.

Reply:
> two sources of different names and also different versions (v2 / v2alpha1), right?

Nope, resources are unique for `(group, namespace, name)`. Version is only used for data format.

Reply:
Given that then I think this PR is built on a false premise. What's not clear to me though is how we do we avoid the deprecation warning messages. My understanding is that because v2alpha1 is marked as deprecated, the K8s informer will automatically issue a warning log, so it's not directly in our control. The only option I see is to add an exception to the CI suite specifically for that warning msg here: https://github.com/cilium/cilium/blob/1293a478675dec3e7ff097dc81283287ed8992f4/test/helpers/cons.go#L270. What this means practically for users is that they will see this warning log if v2alpha1 resources exist and until they migrate over to v2. That sounds reasonable enough to me?